File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # 3 pointers a, b, c 
        # sort the array and shift a and b when they equal the previous value
        # [-4, -1, -1, 0, 1, 2] res = []
        #              a  bc 
        # use 2 pointers to iterate through array
        nums.sort()
        res = []
        for a in range(len(nums)):
            if nums[a] > 0:
                break
            if a > 0 and nums[a] == nums[a-1]:
                continue
            b = a+1
            c = len(nums)-1
            while b < c:
                ssum = nums[a] + nums[b] + nums[c]
                if ssum == 0:
                    res.append([nums[a], nums[b], nums[c]])
                    b += 1
                    c -= 1
                    while b < c and nums[b] == nums[b-1]:
                        b += 1
                elif ssum < 0:
                    b += 1
                else:
                    c -= 1
        return res
            

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();;
        // sort array
        Arrays.sort(nums);
        // iterate through the array and run 2 sum on pair of numbers
        for (int i=0; i<nums.length; i++) {
            if (nums[i] > 0) { // check for positive value
                break;
            }
            if (i > 0 && nums[i] == nums[i-1]) { // check for duplicate a values
                continue;
            }
            int a = nums[i];
            int left = i+1;
            int right = nums.length-1;
            while (left < right) {
                int ssum = a + nums[left] + nums[right];
                if (ssum == 0) {
                    res.add(Arrays.asList(a, nums[left], nums[right]));
                    left++;
                    right--;
                    // check for duplicate b values
                    while (left < right && nums[left] == nums[left-1]) {
                        left++;
                    }
                }
                else if (ssum > 0) {
                    right--;
                } else {
                    left++;
                }
            }
        }
        return res;
    }
}